# W03 — FAISS RAG Pipeline (Fari eldercare / Senpai grounding)

Same architecture pattern as Psychologist Agent's RAG (sentence-transformers + FAISS) and OpenClaw's Knowledge agent (384-dim, L2-normalized). Knowledge base: `knowledge_base.json`, 13 chunks grounded in real public sources (NIA, CDC, WHO, PMC research), each with a title/publisher/URL. Two chunks are flagged as my own extrapolation rather than a direct quote, since I couldn't find a source for those specific points.

Steps: embed → FAISS index → retrieval eval on a held-out query set (`eval_queries.json`, written after the index and not used to tune it) → top-k hit rate.

In [1]:
import json
from pathlib import Path

import numpy as np
import pandas as pd
import faiss
from sentence_transformers import SentenceTransformer

NOTEBOOK_DIR = Path.cwd()
SEED = 42
np.random.seed(SEED)

EMBED_MODEL_NAME = "sentence-transformers/all-MiniLM-L6-v2"  # 384-dim, matches OpenClaw's embedding size

/Users/yihangsun/Desktop/2026找工/8 bit/inGenInternshipDataSources/yihang-ingen-ml-nn-intern/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


## 1. Load knowledge base

In [2]:
with open(NOTEBOOK_DIR / "knowledge_base.json") as f:
    kb = json.load(f)

chunks = kb["chunks"]
chunk_ids = [c["chunk_id"] for c in chunks]
chunk_texts = [c["text"] for c in chunks]
print(f"Loaded {len(chunks)} chunks")
pd.DataFrame([{"chunk_id": c["chunk_id"], "topic": c["topic"], "source": c["source"]["title"]} for c in chunks])

Loaded 13 chunks


,chunk_id,topic,source
0,dementia_communication,Communicating with someone who has dementia or...,Communicating With Someone Who Has Alzheimer's...
1,disorientation_dementia,Responding to confusion or disorientation,Communicating With Someone Who Has Alzheimer's...
2,elderspeak_avoidance,Avoiding patronizing tone ('elderspeak'),Elderspeak to Resident Dementia Patients Incre...
3,loneliness_response,Responding to loneliness and social isolation,Loneliness and Social Isolation — Tips for Sta...
4,family_connection,Facilitating family connection,Loneliness and Social Isolation — Tips for Sta...
5,grief_support,Grief and mourning support,Grief and mourning / Coping With Grief and Loss
6,medication_reminders,Medication adherence reminders,"Attitudes, Beliefs, and Cost-Related Medicatio..."
7,crisis_recognition,Recognizing mental health risk in older adults,Mental health of older adults
8,escalation_handoff,When to hand off to a human / specialized care,Mental health of older adults
9,fall_prevention_context,Fall risk and prevention framework,STEADI — Older Adult Fall Prevention


## 2. Embedding generation

In [3]:
embedder = SentenceTransformer(EMBED_MODEL_NAME)

chunk_embeddings = embedder.encode(chunk_texts, convert_to_numpy=True, show_progress_bar=False).astype("float32")
faiss.normalize_L2(chunk_embeddings)  # so inner product = cosine similarity

print(f"Embedding shape: {chunk_embeddings.shape}")

/Users/yihangsun/Desktop/2026找工/8 bit/inGenInternshipDataSources/yihang-ingen-ml-nn-intern/.venv/lib/python3.12/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


Embedding shape: (13, 384)


## 3. FAISS indexing

In [4]:
index = faiss.IndexFlatIP(chunk_embeddings.shape[1])
index.add(chunk_embeddings)
print(f"FAISS index: {index.ntotal} vectors, dim={chunk_embeddings.shape[1]}")


def retrieve(query, top_k=5):
    q_emb = embedder.encode([query], convert_to_numpy=True).astype("float32")
    faiss.normalize_L2(q_emb)
    scores, idx = index.search(q_emb, top_k)
    return [{"chunk_id": chunk_ids[i], "topic": chunks[i]["topic"], "score": float(s)} for s, i in zip(scores[0], idx[0])]


# sanity check, not part of the held-out eval set
for r in retrieve("what should I say to someone right after they fell", top_k=3):
    print(f"{r['score']:.3f}  {r['chunk_id']:<28}  {r['topic']}")

FAISS index: 13 vectors, dim=384
0.387  fall_prevention_context       Fall risk and prevention framework
0.330  grief_support                 Grief and mourning support
0.275  dementia_communication        Communicating with someone who has dementia or Alzheimer's


## 4. Retrieval evaluation (held-out query set)

`eval_queries.json` was written after the index, phrased differently from the chunk text, and never used to tune retrieval.

In [5]:
with open(NOTEBOOK_DIR / "eval_queries.json") as f:
    eval_queries = json.load(f)["queries"]

print(f"{len(eval_queries)} held-out queries")

K_VALUES = [1, 3, 5]
rows = []
for q in eval_queries:
    results = retrieve(q["query"], top_k=max(K_VALUES))
    row = {"query": q["query"], "relevant_chunk_id": q["relevant_chunk_id"], "top1_retrieved": results[0]["chunk_id"]}
    for k in K_VALUES:
        row[f"hit@{k}"] = q["relevant_chunk_id"] in [r["chunk_id"] for r in results[:k]]
    rows.append(row)

eval_df = pd.DataFrame(rows)
eval_df

13 held-out queries


,query,relevant_chunk_id,top1_retrieved,hit@1,hit@3,hit@5
0,My resident with Alzheimer's is getting frustr...,dementia_communication,dementia_communication,True,True,True
1,Someone suddenly seems confused about where th...,disorientation_dementia,disorientation_dementia,True,True,True
2,"Why is talking to older adults in a slow, sing...",elderspeak_avoidance,elderspeak_avoidance,True,True,True
3,A resident hasn't heard from anyone in a while...,loneliness_response,family_connection,False,True,True
4,What's a good way to help a resident stay in t...,family_connection,family_connection,True,True,True
5,It's the anniversary of my resident's spouse p...,grief_support,grief_support,True,True,True
6,A resident keeps refusing to confirm she took ...,medication_reminders,medication_reminders,True,True,True
7,What kind of statements from an older adult sh...,crisis_recognition,crisis_recognition,True,True,True
8,When should the system stop trying to handle s...,escalation_handoff,escalation_handoff,True,True,True
9,"How common are falls among elderly people, and...",fall_prevention_context,fall_prevention_context,True,True,True


In [6]:
print("Top-k hit rate:\n")
for k in K_VALUES:
    rate = eval_df[f"hit@{k}"].mean()
    print(f"  hit@{k}: {rate:.3f}  ({eval_df[f'hit@{k}'].sum()}/{len(eval_df)})")

misses = eval_df[~eval_df["hit@3"]]
print(f"\nMissed at hit@3 (n={len(misses)}):")
misses[["query", "relevant_chunk_id", "top1_retrieved"]]

Top-k hit rate:

  hit@1: 0.846  (11/13)
  hit@3: 0.923  (12/13)
  hit@5: 0.923  (12/13)

Missed at hit@3 (n=1):


,query,relevant_chunk_id,top1_retrieved
12,What's a good way to build day-to-day rapport ...,companionship_routine,medication_reminders


## 5. Discussion

13 chunks, 13 queries, hit@1=0.846, hit@3=0.923, hit@5=0.923. Small on purpose, this is about the eval methodology (held-out set, never used to tune the index), not a production-scale knowledge base.

One query never hits at all, even at k=5: "What's a good way to build day-to-day rapport that isn't just task reminders?" is supposed to match `companionship_routine`, but top-1 is `medication_reminders` instead. Makes sense once you look at it: the query literally contains the word "reminders," which pulls it toward the medication chunk on pure lexical overlap, while `companionship_routine` is the one chunk I flagged as my own reasoning rather than a direct source quote, so it's also the vaguest, least concrete chunk in the whole set, worst candidate for embedding similarity. That's a real limitation, not a bug: the chunk that's weakest on grounding is also the one retrieval struggles with most.

The other near-miss, `loneliness_response` landing at hit@3 instead of hit@1 (pulled toward `family_connection` first), is less concerning since both chunks come from the same NIA source and are genuinely about overlapping territory.